curvlinops/kfac.py

```diff
@@ -312,6 +312,90 @@ def _reset_matrix_properties(self):
         self._frobenius_norm = None
 
+    @staticmethod
+    def _left_and_right_multiply(
+        M_joint: Tensor,
+        aaT: FactorType,
+        ggT: FactorType,
+        eigenvalues: Optional[Tensor] = None,
+    ) -> Tensor:
+        """Left and right multiply matrix with Kronecker factors.
+
+        Args:
+            M_joint: Matrix for multiplication.
+            aaT: Input covariance Kronecker factor or its eigenvectors. ``None`` for
+                biases.
+            ggT: Gradient covariance Kronecker factor or its eigenvectors.
+            eigenvalues: Eigenvalues of the (E)KFAC approximation when multiplying with
+                the eigendecomposition of the KFAC approximation. ``None`` for the
+                non-decomposed KFAC approximation. Defaults to ``None``.
+
+        Returns:
+            Matrix-multiplication result ``KFAC @ M_joint``.
+        """
+        if eigenvalues is None:
+            M_joint = einsum(ggT, M_joint, aaT, "i j, j k v, k l -> i l v")
+        else:
+            # Perform preconditioning in KFE, e.g. see equation (21) in
+            # https://arxiv.org/abs/2308.03296.
+            aaT_eigvecs = aaT
+            ggT_eigvecs = ggT
+            # Transform in eigenbasis.
+            M_joint = einsum(
+                ggT_eigvecs, M_joint, aaT_eigvecs, "i j, i k v, k l -> j l v"
+            )
+            # Multiply (broadcasted) by eigenvalues.
+            M_joint.mul_(eigenvalues.unsqueeze(-1))
+            # Transform back to standard basis.
+            M_joint = einsum(
+                ggT_eigvecs, M_joint, aaT_eigvecs, "i j, j k v, l k -> i l v"
+            )
+        return M_joint
+
+    @staticmethod
+    def _separate_left_and_right_multiply(
+        KM: List[Tensor],
+        M: Tensor,
+        param_pos: Dict[str, int],
+        aaT: FactorType,
+        ggT: FactorType,
+        eigenvalues: Optional[Tensor] = None,
+    ) -> Tensor:
+        """Multiply matrix with Kronecker factors for separated weight and bias.
+
+        Args:
+            KM: List to write the matrix-multiplication result to.
+            M: Matrix for multiplication.
+            param_pos: Dictionary with positions of the weight and bias parameters.
+            aaT: Input covariance Kronecker factor or its eigenvectors. ``None`` for
+                biases.
+            ggT: Gradient covariance Kronecker factor or its eigenvectors.
+            eigenvalues: Eigenvalues of the (E)KFAC approximation when multiplying with
+                the eigendecomposition of the KFAC approximation. ``None`` for the
+                non-decomposed KFAC approximation. Defaults to ``None``.
+        """
+        for p_name, pos in param_pos.items():
+            # for weights we need to multiply from the right with aaT
+            # for weights and biases we need to multiply from the left with ggT
+            if p_name == "weight":
+                M_w = rearrange(M[pos], "m ... v -> m (...) v")
```

Comment:
`M[pos]` hints that the type annotation of `M` should also be `List[Tensor]` (but it is `Tensor` at the moment). Can you also update the docstring accordingly?